File: backend/models/workflows.py
from uuid import UUID
from datetime import datetime
from typing import Optional, Any
from pydantic import BaseModel


# ─────────────────────────────────────────────────────────────────────────────
# Workflow Definition
# ─────────────────────────────────────────────────────────────────────────────

class CreateWorkflowStageRequest(BaseModel):
    name: str
    stage_order: int
    assigned_role: Optional[str] = None    # staff | manager | admin | super_admin | vendor
    assigned_user_id: Optional[UUID] = None
    action_type: str                        # review | approve | fill_form | sign | create_task | create_issue | create_incident | notify | wait
    form_template_id: Optional[UUID] = None
    is_final: bool = False
    config: Optional[dict] = None          # system stage config (title, priority, hours, etc.)
    sla_hours: Optional[int] = None


class UpdateWorkflowStageRequest(BaseModel):
    name: Optional[str] = None
    stage_order: Optional[int] = None
    assigned_role: Optional[str] = None
    assigned_user_id: Optional[UUID] = None
    action_type: Optional[str] = None
    form_template_id: Optional[UUID] = None
    is_final: Optional[bool] = None
    config: Optional[dict] = None
    sla_hours: Optional[int] = None


class ReorderStagesRequest(BaseModel):
    stages: list[dict]  # [{id: str, stage_order: int}]


class CreateRoutingRuleRequest(BaseModel):
    from_stage_id: UUID
    to_stage_id: UUID
    condition_type: str   # score_below | score_above | field_failed | field_value_equals | always | approved | rejected | priority_equals | role_equals | sla_breached
    condition_field_id: Optional[UUID] = None
    condition_value: Optional[str] = None
    priority: int = 0
    label: Optional[str] = None


class UpdateRoutingRuleRequest(BaseModel):
    condition_type: Optional[str] = None
    condition_value: Optional[str] = None
    priority: Optional[int] = None
    label: Optional[str] = None


class CreateWorkflowDefinitionRequest(BaseModel):
    name: str
    trigger_type: str = "manual"            # manual | audit_submitted | issue_created | incident_created | scheduled | form_submitted
    trigger_config: Optional[dict] = None
    form_template_id: Optional[UUID] = None  # required when trigger_type = audit_submitted | form_submitted
    is_active: bool = True
    stages: list[CreateWorkflowStageRequest] = []


class UpdateWorkflowDefinitionRequest(BaseModel):
    name: Optional[str] = None
    is_active: Optional[bool] = None
    trigger_type: Optional[str] = None
    trigger_config: Optional[dict] = None


# ─────────────────────────────────────────────────────────────────────────────
# Workflow Instance
# ─────────────────────────────────────────────────────────────────────────────

class TriggerWorkflowRequest(BaseModel):
    definition_id: UUID
    source_type: str = "manual"           # manual | issue | incident | audit
    source_id: Optional[UUID] = None
    location_id: Optional[UUID] = None


class CancelWorkflowRequest(BaseModel):
    reason: str


class WorkflowStageInstanceResponse(BaseModel):
    id: UUID
    stage_id: UUID
    assigned_to: Optional[UUID] = None
    status: str
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    due_at: Optional[datetime] = None
    comment: Optional[str] = None
    form_submission_id: Optional[UUID] = None
    spawned_task_id: Optional[UUID] = None
    spawned_issue_id: Optional[UUID] = None
    spawned_incident_id: Optional[UUID] = None
    stage_name: Optional[str] = None
    stage_order: Optional[int] = None
    action_type: Optional[str] = None
    is_final: Optional[bool] = None


class WorkflowInstanceResponse(BaseModel):
    id: UUID
    workflow_definition_id: UUID
    submission_id: Optional[UUID] = None
    source_type: Optional[str] = None
    source_id: Optional[UUID] = None
    organisation_id: UUID
    status: str
    current_stage_id: Optional[UUID] = None
    completed_at: Optional[datetime] = None
    created_at: datetime
    workflow_name: Optional[str] = None
    stage_instances: list[WorkflowStageInstanceResponse] = []


class ApproveStageRequest(BaseModel):
    comment: Optional[str] = None


class RejectStageRequest(BaseModel):
    comment: str    # required on rejection


class SubmitFormForStageRequest(BaseModel):
    responses: list[dict]    # [{field_id, value}]

File: frontend/services/workflows.ts
import { apiFetch } from "@/services/api/client";
import { getClientToken } from "@/lib/auth";

// ─── Types ──────────────────────────────────────────────────────────────────

export interface WorkflowStage {
  id: string;
  workflow_definition_id: string;
  name: string;
  stage_order: number;
  action_type: string;
  assigned_role?: string | null;
  assigned_user_id?: string | null;
  form_template_id?: string | null;
  is_final: boolean;
  config?: Record<string, unknown> | null;
  sla_hours?: number | null;
  is_deleted?: boolean;
}

export interface RoutingRule {
  id: string;
  workflow_definition_id: string;
  from_stage_id: string;
  to_stage_id: string;
  condition_type: string;
  condition_field_id?: string | null;
  condition_value?: string | null;
  priority: number;
  label?: string | null;
  is_deleted?: boolean;
}

export interface WorkflowDefinition {
  id: string;
  name: string;
  trigger_type: string;
  trigger_config?: Record<string, unknown> | null;
  form_template_id?: string | null;
  trigger_form_template_id?: string | null;
  trigger_issue_category_id?: string | null;
  trigger_conditions?: Record<string, unknown> | null;
  is_active: boolean;
  created_at: string;
  updated_at?: string;
  workflow_stages?: WorkflowStage[];
  workflow_routing_rules?: RoutingRule[];
}

export interface WorkflowStageInstance {
  id: string;
  stage_id: string;
  workflow_instance_id?: string | null;
  assigned_to?: string | null;
  status: string;
  started_at?: string | null;
  completed_at?: string | null;
  due_at?: string | null;
  comment?: string | null;
  form_submission_id?: string | null;
  spawned_task_id?: string | null;
  spawned_issue_id?: string | null;
  spawned_incident_id?: string | null;
  workflow_stages?: {
    name: string;
    action_type: string;
    stage_order: number;
    is_final?: boolean;
    form_template_id?: string | null;
    config?: Record<string, unknown> | null;
    form_templates?: { title: string } | null;
  } | null;
  review_submission_id?: string | null;
  stage_history?: {
    id: string;
    stage_name: string | null;
    action_type: string | null;
    stage_order: number | null;
    status: string;
    completed_at: string | null;
    comment: string | null;
    completed_by: string | null;
    form_submission_id: string | null;
  }[] | null;
  workflow_instances?: {
    id: string;
    status: string;
    source_type?: string | null;
    source_id?: string | null;
    organisation_id: string;
    workflow_definitions?: {
      name: string;
    } | null;
  } | null;
}

export interface WorkflowInstance {
  id: string;
  workflow_definition_id: string;
  organisation_id: string;
  source_type?: string | null;
  source_id?: string | null;
  submission_id?: string | null;
  location_id?: string | null;
  triggered_by?: string | null;
  status: string;
  current_stage_id?: string | null;
  cancelled_reason?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
  created_at: string;
  workflow_definitions?: { name: string; trigger_type: string } | null;
  workflow_stages?: { name: string; action_type: string } | null;
  workflow_stage_instances?: WorkflowStageInstance[];
}

// ─── Definitions ─────────────────────────────────────────────────────────────

export async function listWorkflowDefinitions(): Promise<WorkflowDefinition[]> {
  return apiFetch("/api/v1/workflows/definitions");
}

export async function getWorkflowDefinition(id: string): Promise<WorkflowDefinition> {
  return apiFetch(`/api/v1/workflows/definitions/${id}`);
}

export async function createWorkflowDefinition(body: {
  name: string;
  trigger_type: string;
  form_template_id?: string | null;
  trigger_config?: Record<string, unknown> | null;
  is_active?: boolean;
  stages?: Partial<WorkflowStage>[];
}): Promise<WorkflowDefinition> {
  return apiFetch("/api/v1/workflows/definitions", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export async function updateWorkflowDefinition(
  id: string,
  body: Partial<Pick<WorkflowDefinition, "name" | "is_active" | "trigger_type" | "trigger_config">>
): Promise<{ success: boolean }> {
  return apiFetch(`/api/v1/workflows/definitions/${id}`, {
    method: "PUT",
    body: JSON.stringify(body),
  });
}

export async function deleteWorkflowDefinition(id: string): Promise<{ success: boolean }> {
  return apiFetch(`/api/v1/workflows/definitions/${id}`, { method: "DELETE" });
}

export async function duplicateWorkflowDefinition(id: string): Promise<WorkflowDefinition> {
  return apiFetch(`/api/v1/workflows/definitions/${id}/duplicate`, { method: "POST" });
}

// ─── Stages ───────────────────────────────────────────────────────────────────

export async function addStage(
  definitionId: string,
  body: Partial<WorkflowStage>
): Promise<WorkflowStage> {
  return apiFetch(`/api/v1/workflows/definitions/${definitionId}/stages`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export async function updateStage(
  definitionId: string,
  stageId: string,
  body: Partial<WorkflowStage>
): Promise<{ success: boolean }> {
  return apiFetch(`/api/v1/workflows/definitions/${definitionId}/stages/${stageId}`, {
    method: "PUT",
    body: JSON.stringify(body),
  });
}

export async function deleteStage(
  definitionId: string,
  stageId: string
): Promise<{ success: boolean }> {
  return apiFetch(`/api/v1/workflows/definitions/${definitionId}/stages/${stageId}`, {
    method: "DELETE",
  });
}

export async function reorderStages(
  definitionId: string,
  stages: { id: string; stage_order: number }[]
): Promise<{ success: boolean }> {
  return apiFetch(`/api/v1/workflows/definitions/${definitionId}/stages/reorder`, {
    method: "PUT",
    body: JSON.stringify({ stages }),
  });
}

// ─── Routing Rules ────────────────────────────────────────────────────────────

export async function addRoutingRule(
  definitionId: string,
  body: Omit<RoutingRule, "id" | "workflow_definition_id">
): Promise<RoutingRule> {
  return apiFetch(`/api/v1/workflows/definitions/${definitionId}/routing-rules`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export async function updateRoutingRule(
  definitionId: string,
  ruleId: string,
  body: Partial<Pick<RoutingRule, "condition_type" | "condition_value" | "priority" | "label">>
): Promise<{ success: boolean }> {
  return apiFetch(`/api/v1/workflows/definitions/${definitionId}/rules/${ruleId}`, {
    method: "PUT",
    body: JSON.stringify(body),
  });
}

export async function deleteRoutingRule(
  definitionId: string,
  ruleId: string
): Promise<{ success: boolean }> {
  return apiFetch(`/api/v1/workflows/definitions/${definitionId}/rules/${ruleId}`, {
    method: "DELETE",
  });
}

// ─── Instances ────────────────────────────────────────────────────────────────

export async function listInstances(filters?: {
  status?: string;
  location_id?: string;
  definition_id?: string;
  from?: string;
  to?: string;
  page?: number;
  page_size?: number;
  my_team?: boolean;
}): Promise<WorkflowInstance[]> {
  const params = new URLSearchParams();
  if (filters?.status) params.set("status", filters.status);
  if (filters?.location_id) params.set("location_id", filters.location_id);
  if (filters?.definition_id) params.set("definition_id", filters.definition_id);
  if (filters?.from) params.set("from", filters.from);
  if (filters?.to) params.set("to", filters.to);
  if (filters?.page) params.set("page", String(filters.page));
  if (filters?.page_size) params.set("page_size", String(filters.page_size));
  if (filters?.my_team) params.set("my_team", "true");
  return apiFetch(`/api/v1/workflows/instances?${params}`);
}

export async function getInstanceDetail(id: string): Promise<WorkflowInstance> {
  return apiFetch(`/api/v1/workflows/instances/${id}`);
}

export async function triggerWorkflow(body: {
  definition_id: string;
  source_type?: string;
  source_id?: string;
  location_id?: string;
}): Promise<WorkflowInstance> {
  return apiFetch("/api/v1/workflows/instances", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export async function cancelInstance(id: string, reason: string): Promise<{ success: boolean }> {
  return apiFetch(`/api/v1/workflows/instances/${id}/cancel`, {
    method: "POST",
    body: JSON.stringify({ reason }),
  });
}

export async function approveStage(
  instanceId: string,
  stageInstanceId: string,
  comment?: string
): Promise<{ status: string }> {
  return apiFetch(`/api/v1/workflows/instances/${instanceId}/stages/${stageInstanceId}/approve`, {
    method: "POST",
    body: JSON.stringify({ comment }),
  });
}

export async function rejectStage(
  instanceId: string,
  stageInstanceId: string,
  comment: string
): Promise<{ status: string }> {
  return apiFetch(`/api/v1/workflows/instances/${instanceId}/stages/${stageInstanceId}/reject`, {
    method: "POST",
    body: JSON.stringify({ comment }),
  });
}

export async function getMyWorkflowTasks(): Promise<WorkflowStageInstance[]> {
  return apiFetch("/api/v1/workflows/instances/my-tasks");
}

export async function getStageInstance(
  instanceId: string,
  stageInstanceId: string
): Promise<WorkflowStageInstance> {
  return apiFetch(`/api/v1/workflows/instances/${instanceId}/stages/${stageInstanceId}`);
}

export async function submitFormForStage(
  instanceId: string,
  stageInstanceId: string,
  responses: { field_id: string; value: string; comment?: string }[]
): Promise<{ success: boolean; form_submission_id: string }> {
  return apiFetch(`/api/v1/workflows/instances/${instanceId}/stages/${stageInstanceId}/submit-form`, {
    method: "POST",
    body: JSON.stringify({ responses }),
  });
}

// ─── AI Generation ────────────────────────────────────────────────────────────

export interface GeneratedWorkflowStage {
  name: string;
  action_type: string;
  assigned_role?: string | null;
  sla_hours?: number | null;
  is_final: boolean;
  config?: Record<string, unknown> | null;
}

export interface GeneratedWorkflow {
  name: string;
  trigger_type: string;
  stages: GeneratedWorkflowStage[];
}

export class PublishValidationError extends Error {
  errors: string[];
  constructor(errors: string[]) {
    super("Publish validation failed");
    this.errors = errors;
  }
}

export async function publishWorkflow(id: string): Promise<{ success: boolean }> {
  // Can't use apiFetch directly — need to inspect 422 body for validation errors
  const token = getClientToken();
  const API_BASE = process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:8000";

  const res = await fetch(`${API_BASE}/api/v1/workflows/definitions/${id}/publish`, {
    method: "POST",
    headers: { "Content-Type": "application/json", ...(token ? { Authorization: `Bearer ${token}` } : {}) },
  });

  if (res.status === 422) {
    const body = await res.json().catch(() => ({}));
    const errors: string[] = body?.detail?.errors ?? [body?.detail ?? "Validation failed"];
    throw new PublishValidationError(errors);
  }
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error(body?.detail || `HTTP ${res.status}`);
  }
  return res.json();
}

export async function generateWorkflowWithAI(prompt: string): Promise<GeneratedWorkflow> {
  return apiFetch("/api/v1/ai/generate-workflow", {
    method: "POST",
    body: JSON.stringify({ prompt }),
  });
}
